File: gallery/process_counter_ios/ios/Host/ProcessUiBridge.swift
/// SwiftUI refresh hook. Generated `ProcessUiHost.notifyPath` is empty — use
/// `dispatch(path:)` from a future codegen forwarder, or poll `__rangerStateGeneration`.
enum ProcessUiBridge {
    private static var pathListeners: [String: [() -> Void]] = [:]

    static func subscribe(path: String, listener: @escaping () -> Void) -> () -> Void {
        var set = pathListeners[path] ?? []
        set.append(listener)
        pathListeners[path] = set
        let id = set.count - 1
        return {
            var current = pathListeners[path] ?? []
            if id < current.count {
                current.remove(at: id)
            }
            pathListeners[path] = current
        }
    }

    static func dispatch(path: String) {
        pathListeners[path]?.forEach { $0() }
    }
}

File: gallery/process_counter_android/app/host/ProcessUiBridge.kt
/**
 * Android UI refresh hook.
 *
 * Generated [ProcessUiHost.notifyPath] is an empty method (not a Kotlin property),
 * unlike the TS gallery where we assign `host.notifyPath = …`. Until codegen emits
 * a forwarder, call [notifyPath] from host code after actions, or poll
 * [CounterBoardPage.__rangerStateGeneration] (see [CounterBoardController]).
 */
object ProcessUiBridge {
    private val pathListeners = mutableMapOf<String, MutableSet<() -> Unit>>()

    fun subscribePath(path: String, listener: () -> Unit): () -> Unit {
        val set = pathListeners.getOrPut(path) { mutableSetOf() }
        set.add(listener)
        return { set.remove(listener) }
    }

    /** Call from a patched ProcessUiHost or tests when notifyPath is wired. */
    fun dispatchPath(path: String) {
        pathListeners[path]?.forEach { it.invoke() }
    }
}
